e2e/tests/components/logical-expression-builder/logical-expression-builder.e2e-spec.ts

```diff
@@ -0,0 +1,183 @@
+import { LogicalExpressionBuilderPage } from './logical-expression-builder.po.spec';
+import { ExpressionCondition, ExpressionGroup } from '../../../../src/components/logical-expression-builder';
+
+describe('Logical Expression Builder', () => {
+    let page: LogicalExpressionBuilderPage.Page;
+
+    beforeEach(async () => {
+        page = new LogicalExpressionBuilderPage.Page();
+        await page.getPage();
+    });
+
+    it('should have correct initial state', async () => {
+        const btn = await page.getEmptyExpressionButton();
+
+        expect(await page.getExpressionObject()).toEqual(null);
+        expect(await page.getConditionRowCount()).toEqual(0);
+        expect(await page.getGroupRowCount()).toEqual(0);
+        expect(await page.getValid()).toBeTruthy();
+        expect(btn).toBeDefined();
+        expect(await btn.getText()).toEqual('ADD CONDITION');
+    });
+
+    it('should display one row when expression is just a condition', async () => {
+        await page.setOneCondition();
```

Comment:
For an e2e (end-to-end) test case, you would build the expression as a user would. A user would not click a button to pre-fill the component, they would click the "add condition" button, click the dropdown, select a field type, and so on. What you have implemented here are more suitable as karma tests, because user interaction is not considered.

Since this component is replacing an existing AngularJS component, I assume that the use cases are well understood. Those use cases should convert quite readily into e2e test cases. If you would like some advice on designing e2e test cases, we can discuss it.

Reply:
Okay, that makes sense. I think discussing this is a good idea since honestly, I haven't implemented E2E tests before this.